persistent-postgresql/test/UpsertWhere.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE DataKinds, FlexibleInstances, MultiParamTypeClasses, ExistentialQuantification #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
+module UpsertWhere where
+
+import Data.List              (sort)
+
+import Database.Persist.Postgresql
+import PgInit
+
+share [mkPersist sqlSettings, mkMigrate "upsertWhereMigrate"] [persistLowerCase|
+  Item
+     name        Text sqltype=varchar(80)
+     description Text
+     price       Double Maybe
+     quantity    Int Maybe
+
+     Primary name
+     deriving Eq Show Ord
+
+|]
+
+specs :: Spec
+specs = describe "UpsertWhere" $ do
+  let item1 = Item "item1" "" (Just 3) Nothing
+      item2 = Item "item2" "hello world" Nothing (Just 2)
+      items = [item1, item2]
+  describe "upsertWhere" $ do
+    it "inserts appropriately" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      upsertWhere item1 [ItemDescription =. "i am item 1"] []
+      Just item <- get (ItemKey "item1")
+      item @== item1
+
+    it "performs only updates given if record already exists" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      let newDescription = "I am a new description"
+      insert_ item1
+      upsertWhere
+        (Item "item1" "i am inserted description" (Just 1) (Just 2))
+        [ItemDescription =. newDescription] 
+        []
+      Just item <- get (ItemKey "item1")
+      item @== item1 { itemDescription = newDescription }
+
+  describe "upsertManyWhere" $ do
+    it "inserts fresh records" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      let newItem = Item "item3" "fresh" Nothing Nothing
+      upsertManyWhere
+        (newItem : items)
+        [copyField ItemDescription]
+        []
+        []
+      dbItems <- map entityVal <$> selectList [] []
+      sort dbItems @== sort (newItem : items)
+    it "updates existing records" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      upsertManyWhere
+        items
+        []
+        [ItemQuantity +=. Just 1]
+        []
+    -- TODO: this test doesn't pass
+    it "only copies passing values" $ runConnAssert $ do
```

Comment:
this test fails due to an error like this
```
UpsertWhere.upsertManyWhere only copies passing values
       uncaught exception: FormatError
       FormatError {fmtMessage = "8 single '?' characters, but 9 parameters", fmtQuery = "INSERT INTO \"item\"(\"name\", \"description\", \"price\", \"quantity\") VALUES (?, ?, ?, ?), (?, ?, ?, ?) ON CONFLICT (\"name\") DO UPDATE SET \"price\"=EXCLUDED.\"price\", \"quantity\"=EXCLUDED.\"quantity\"", fmtParams = ["item1","","6.0","0","item2","hello world","null","0","0"]}
```

Where in my formatting am I missing the value that I need?

Reply:
Let's write some tests for the `mkBulkUpsertQuery` to see what's getting generated here.

I think we'd need to see the generated SQL to get an easy sense of what's going on here.

Reply:
yeah I should've included it; here's the generated SQL
```
INSERT INTO "item"("name", "description", "price", "quantity") VALUES (?, ?, ?, ?), (?, ?, ?, ?) ON CONFLICT ("name") DO UPDATE SET "price"=EXCLUDED."price", "quantity"=EXCLUDED."quantity"
```

Reply:
Can you also dump the `[PersistValue]` that we're generating? The error says that we're giving it `9` parameters.

Reply:
```
[PersistText "item1",PersistText "",PersistDouble 6.0,PersistInt64 0,PersistText "item2",PersistText "hello world",PersistNull,PersistInt64 0,PersistInt64 0]
```

Reply:
So we have `PersistInt64 0` in there twice. We should have the first one (since that't the `Just 0` from the `quantity`), but the second one is spurious.

The `DO UPDATE` clause mentions price and quantity, and we've got `copyField ItemPrice` which is fine and `copyUnlessEq ItmeQuantity (Just 0)`. But the generated SQL has `DO UPDATE SET "price"=EXCLUDED."price", "quantity"=EXCLUDED."quantity"` - these are both `CopyField`-looking to me. Looking at the implementation...